dlup/experimental_backends/_tifffile.py

```diff
@@ -0,0 +1,208 @@
+# coding=utf-8
+# Copyright (c) dlup contributors
+import os
+from typing import Any, Dict, Tuple
+
+import numpy as np
+import PIL.Image
+import tifffile
+
+from dlup.experimental_backends.common import AbstractSlideBackend, numpy_to_pil
+
+
+def open_slide(filename: os.PathLike) -> "TifffileSlide":
+    """
+    Read slide with tifffile.
+
+    Parameters
+    ----------
+    filename : PathLike
+        Path to image.
+    """
+    return TifffileSlide(filename)
+
+
+def get_tile(page: tifffile.TiffPage, coordinates: Tuple[Any, ...], size: Tuple[Any, ...]) -> np.ndarray:
+    """Extract a crop from a TIFF image file directory (IFD).
+
+    Only the tiles englobing the crop area are loaded and not the whole page.
+    This is useful for large Whole slide images that can't fit int RAM.
+
+    Code obtained from [1].
+
+    Parameters
+    ----------
+    page : TiffPage
+        TIFF image file directory (IFD) from which the crop must be extracted.
+    coordinates: (int, int)
+        Coordinates of the top left and right corner corner of the desired crop.
+    size: (int, int)
+        Desired crop height and width.
+
+    References
+    ----------
+    .. [1] https://gist.github.com/rfezzani/b4b8852c5a48a901c1e94e09feb34743
+
+    Returns
+    -------
+    out : ndarray of shape (imagedepth, h, w, sampleperpixel)
+        Extracted crop.
+
+    """
+    y0, x0 = coordinates
+    w, h = size
+
+    if not page.is_tiled:
+        raise ValueError("Input page must be tiled.")
+
+    image_width = page.imagewidth
+    image_height = page.imagelength
+
+    if h < 1 or w < 1:
+        raise ValueError("h and w must be strictly positive.")
+
+    if y0 < 0 or x0 < 0 or y0 + h > image_height or x0 + w > image_width:
+        raise ValueError("Requested crop area is out of image bounds.")
+
+    tile_width, tile_height = page.tilewidth, page.tilelength
+    y1, x1 = y0 + h, x0 + w
+
+    tile_y0, tile_x0 = y0 // tile_height, x0 // tile_width
+    tile_y1, tile_x1 = np.ceil([y1 / tile_height, x1 / tile_width]).astype(int)
+
+    tile_per_line = int(np.ceil(image_width / tile_width))
+
+    out = np.empty(
+        (page.imagedepth, (tile_y1 - tile_y0) * tile_height, (tile_x1 - tile_x0) * tile_width, page.samplesperpixel),
+        dtype=page.dtype,
+    )
+
+    fh = page.parent.filehandle
+
+    jpeg_tables = page.tags.get("JPEGTables", None)
```

Comment:
Could you explain in layman terms what are you doing here? Just for future readers.

Reply:
I took it from the link in the comment, It's the author providing this for us.